### DISCUSSION TITLE
Permissions and storing lookup tables

### DISCUSSION THREAD

**Discussion by pstaabp** [openwebwork/webwork3#7] [Q&A]
We had a good discussion about permissions moving forward.   It was clear that we need to move away from the numerical/hierarchical form we have now and in same way move to a lookup table for each role.  I have some notes/ideas [here](../blob/main/docs/permissions-handling.md).  

There was discussion about storing the lookup table in the database, but I was wondering if that is overkill since once a course is set, the roles probably won't change.  I always feel that databases are most useful for persistent storage that are changed as a user interacts with a website.  

Since the lookup table will need to be accessed for basically any functionality, the question of where is it stored is one of speed.  It seems like a few options are

1. in the database
2. in config files like current situation of loading the CourseEnvironment
3. in persistent memory (does mojolicious have anything like this)

This may be a broader question in that how all course settings are stored and the answer may be some testing and benchmarking.

**Comment by dlglin** [comment]
One issue that occurred to me with storing configuration in files is that it is more work to build a UI to edit it.  If a bunch of config is stored in a file, then the only way to update a value is to read the file, make the change, then write the whole file.  Storing things in the DB has the advantage of being able to modify one entry without having to look at anything else.
